### ISSUE TITLE
Failed to load the line-diff-details package

### ISSUE THREAD

**Issue by mattdiamond** [jakesankey/line-diff-details#45]
**Atom**: 1.21.0 x64
**Electron**: 1.6.9
**OS**: Mac OS X 10.12.6
**Thrown From**: [line-diff-details](https://github.com/jakesankey/line-diff-details) package 1.9.0

### Stack Trace

Failed to load the line-diff-details package

```
At ENOENT, node_modules/typescript/bin/lib.d.ts not found in /Applications/Atom.app/Contents/Resources/app.asar

Error: ENOENT, node_modules/typescript/bin/lib.d.ts not found in /Applications/Atom.app/Contents/Resources/app.asar
    at notFoundError (ELECTRON_ASAR.js:115:19)
    at Object.fs.readFileSync (ELECTRON_ASAR.js:514:9)
    at TypeScriptSimple.createService (/Applications/Atom.app/Contents/Resources/app/node_modules/typescript-simple/index.js:71:73)
    at TypeScriptSimple.compile (/Applications/Atom.app/Contents/Resources/app/node_modules/typescript-simple/index.js:60:43)
    at Object.exports.compile (/Applications/Atom.app/Contents/Resources/app/src/typescript.js:54:53)
    at Object.compile (/Applications/Atom.app/Contents/Resources/app/src/package-transpilation-registry.js:79:33)
    at compileFileAtPath (/Applications/Atom.app/Contents/Resources/app/src/compile-cache.js:97:37)
    at Object.value [as .ts] (/Applications/Atom.app/Contents/Resources/app/src/compile-cache.js:238:26)
    at Module.load (module.js:488:32)
    at tryModuleLoad (module.js:447:12)
    at Function.Module._load (module.js:439:3)
    at Module.require (/app.asar/static/index.js:47:45)
    at require (internal/module.js:20:19)
    at customRequire (/Applications/Atom.app/Contents/Resources/app/static/<embedded>:96:26)
    at Package.module.exports.Package.requireMainModule (/Applications/Atom.app/Contents/Resources/app/src/package.js:903:35)
    at /Applications/Atom.app/Contents/Resources/app/src/package.js:158:34
    at Package.module.exports.Package.measure (/Applications/Atom.app/Contents/Resources/app/src/package.js:99:21)
    at Package.module.exports.Package.load (/Applications/Atom.app/Contents/Resources/app/src/package.js:144:18)
    at PackageManager.loadAvailablePackage (/Applications/Atom.app/Contents/Resources/app/src/package-manager.js:617:16)
    at config.transact (/Applications/Atom.app/Contents/Resources/app/src/package-manager.js:530:20)
    at Config.module.exports.Config.transact (/Applications/Atom.app/Contents/Resources/app/src/config.js:342:22)
    at PackageManager.loadPackages (/Applications/Atom.app/Contents/Resources/app/src/package-manager.js:528:23)
    at /Applications/Atom.app/Contents/Resources/app/src/atom-environment.js:821:34
```

### Commands

```
  8x -0:11.3.0 core:copy (atom-notification.fatal.icon.icon-bug.native-key-bindings.has-detail.has-close.has-stack)
```

### Non-Core Packages

```
line-diff-details 1.9.0 
```

**Comment by mattdiamond** [comment]
Seeing something similar here:

https://github.com/smhxx/atom-ts-transpiler/issues/8#issuecomment-334346709

**Comment by smhxx** [comment]
I'm able to repro this on 1.21.0 as well. Haven't checked on previous versions, but from what we've deduced, this has to do with the built-in TypeScript support that Atom has, which utilizes `typescript-simple =1.0.0`, which in turn depends on `typescript ~1.4.1`. The location of the core library definitions was moved from `bin/` to `lib/` somewhere between TypeScript 1.4 and 1.5, which means that `typescript-simple` is unable to find the file (lib.d.ts) when installed with the most recent version of TypeScript matching `~1.4.1`. In other words, Atom's built-in TypeScript support is essentially broken, and when it sees `lib/line-diff-details.ts`, it tries to transpile it, but the definitions it needs aren't where it expects them, and boom, error.

There are probably a few equally-valid ways of fixing this:
* Switch `lib/line-diff-details.ts` from TypeScript to JavaScript, optionally with a `"use babel";` directive to enable newer JavaScript features
* Compile the TypeScript *prior* to publishing with apm, and set the "main" file in package.json to the compiled JavaScript output
* Continue to publish the package in TypeScript, but specify a custom transpiler to handle the TS->JS conversion on install/first run by the user (full disclosure: The only one of these that currently exists for TypeScript is the one I wrote, which @mattdiamond linked above, available as `atom-ts-transpiler` on npm.)

Just as a fore-warning if you decide to go with the third approach, `atom-ts-transpiler` is theoretically stable, but still very new, and I'm currently working on getting it field-tested with the help of a few collaborators. As I've dealt with this problem before, (hence writing a package to work around the issue,) I'm around to help out if you need any assistance. 😃

**Comment by smhxx** [comment]
(btw, I've just checked and confirmed that `line-diff-details` does load and run with no issues using `atom-ts-transpiler 1.2.0` and `typescript 2.5.3`, if that's a route you end up wanting to go) 👍

**Comment by robiXxu** [comment]
Also in Windows 10.

**Comment by exside** [comment]
+1

**Comment by mattdiamond** [comment]
> In other words, Atom's built-in TypeScript support is essentially broken

Is there an open issue or pull request to resolve this?

**Comment by smhxx** [comment]
I'm not aware of one, but I haven't checked. It's a pretty simple fix since the bug is just due to a backwards-incompatible change being pushed out in a patch version update (although I don't think the Typescript maintainers realized that `typescript-simple` was relying directly on the lib.d.ts file, so they wouldn't have known moving it would break anything.)

I think it should just require bumping the dependency on `typescript-simple` from 1.0.0 (which is now broken) to any version that uses TypeScript 1.5 or later.

**Comment by mberrueta** [comment]
+1
* **Atom:**       1.22.1
* **Atom-Shell:** 
* **OS:**         Mac OS X 10.13.1
* **Misc**
    * apm  1.18.8
    * npm  3.10.10
    * node 6.9.5 x64
    * python 2.7.10
    * git 2.13.6

**Comment by smhxx** [comment]
Since this problem appears to be ongoing, I suppose it's worth mentioning the official response from one of the Atom devs at atom/atom#15909 (thanks to @mattdiamond for opening that issue)

**Comment by capaj** [comment]
@smhxx  I just ran into this bug this morning. It sure would be nice to get this fixed.

I am on ubuntu 17.10 atom 1.23.3 x64

**Comment by cdbattags** [comment]
This still ain't fixed? 😜

What can I do to help? Are we waiting on another repo to get updated? Can we enact a temp fix and repair back to the proper Atom way when the dep lib gets patched?

**Comment by sebgrizzly** [comment]
hoping for some good news soon. Exactly the package i was lookng for, but during install I end up here. :see_no_evil:

**Comment by smhxx** [comment]
There is a PR open which fixes this bug (#46) but it has changes requested and the author has not updated it yet. If needed, I can open a separate PR so that we can get this package working again, it's been quite a while since it last worked.

**Comment by emanuelegissi** [comment]
Replacing line-diff-details with line-diff-details-plus solves the problem.
